### ISSUE TITLE
cannot query on an emmbedded-list (long) with the operator 'between x and y' 

### ISSUE THREAD

**Issue by T55134** [orientechnologies/orientdb#7373]
### OS: Windows 10  

I cannot query on an emmbedded-list (long) with the operator 'between x and y' and not with '>=x and <=y' (>= alone - works, <= alone - works) - the index is notUnique.

## Expected behavior  
if a record contains the value [2,3,4] in an 'emmbedded-list (long)' field called 'prop', I should get this record when perform a query serach with the condition like: 'prop between 1 and 4', 'prop between 2 and 3'

## Actual behavior  
an exception is thrown in orient server:

> Error on using index EN70.EP70_4 in query 'SELECT * FROM EN70 WHERE (EP70_4 BETWEEN 1 AND 3 )'. Probably you need to rebuild indexes. Now executing query using cluster scan java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.lang.Long at com.orientechnologies.common.serialization.types.OLongSerializer.preprocess(OLongSerializer.java:36) at...

## Steps to reproduce  

1.   I have a vertex called: EN70 with a property: EP70_4 from type 'embedded list = LONG'.
2.   I added a record to this vertex with the value of [2,3,4] in this 'EP70_4' property.
3.   The index on this property: Not_Unique (sbtree). [must support range queries]
4.   SELECT * FROM EN70 WHERE (EP70_4 >= 1) is working and returns my record.
5.   SELECT * FROM EN70 WHERE (EP70_4 <= 3) is working and returns my record.
6.   The operator "Between" is not woking and throw exception: SELECT * FROM EN70 WHERE (EP70_4 BETWEEN 1 AND 3 )
7.   I tried to simulate a "between" operator: SELECT * FROM EN70 WHERE ((EP70_4 >= 1) AND (EP70_4 <= 3)) IS NOT WORKING but it is not working.

both sections 6 and 7 throws an exception:
Error on using index EN70.EP70_4 in query 'SELECT * FROM EN70 WHERE (EP70_4 BETWEEN 1 AND 3 )'. Probably you need to rebuild indexes. Now executing query using cluster scan java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.lang.Long at com.orientechnologies.common.serialization.types.OLongSerializer.preprocess(OLongSerializer.java:36) at...

See also:
[http://stackoverflow.com/questions/43602699/using-beteen-operator-on-a-long-embedded-list-property-declared-in-orientdb-2](url)

**Comment by luigidellaquila** [comment]
Hi @T55134 

I'm working to a fix for this, it will be in the branch in a few hours.

Anyway, IMHO this behavior is intrinsically incorrect. The comparison of a list with a long should always evaluate to `false`, we should define a specific syntax to query elements inside a list. 

Thanks

Luigi

**Comment by T55134** [comment]
Luigi, 

May I ask why do you thing it is wrong to perform a comparison on a list?? 
Let's say I have a vertex for a person with an embedded list for his addresses - if I search for address 'xxx' it looks like this: "WHERE address = 'xxx' " I get this person and it is ok that only one item within the list is a match. (It works fine now, no need change it)

Another example is regarding ranges ('between') - a field is emmbedded list of DATE - Let's say that person vertex has a field like this that contains the dates that this person was using his creditcard. so when I search for "...WHERE ccardUses between '2017-01-01 00:00:00' and '2017-12-31 23:59:59'  " I expect to get this person as a result even ONE of the dates match the condition.

In short - =, >=, <= is working just fine on embedded list but when you combine more than one condition you get an exception and no results: 
- between,
- '>= and <=', 
- in
- etc.

I think that the core of this issue is exactly like this one:
[http://stackoverflow.com/questions/41207737/a-valid-query-syntax-for-orientdb-returns-no-results-without-errors](url)

And there is maybe a different settings I need for the index - I will be happy to know what...

Thanks
Tami

**Comment by luigidellaquila** [comment]
Hi @T55134 

The use case makes a lot of sense, what puzzles me more is the SQL syntax. I'll try to explain what I mean:
```
SELECT FROM V where foo = 10
```
if `foo` is a number, then the query is straight forward

If `foo` is a list, like `[10, 20, 30]` then it becomes more complex: maybe you want to obtain the records where at least one element matches the condition, maybe you want only the record where ALL the elements in the collection match the condition. 
The right syntax for the first case would be

```
SELECT FROM V where foo CONTAINS 10
```
the `=` syntax should succeed only when you pass a list, eg.

```
SELECT FROM V where foo = [10, 20, 30]
```

The same can be applied to `<`, `<=`, `>`, `>=` and so on. IMHO the syntax should be as explicit as you can read it, ie. `10` is greater than `5`, but `10` is not greater than `[5, 6, 7]`, and it's not even greater than `[5, 10, 15]`. This is why I think we should have another operator for this, something like (and I think it actually works)

```
SELECT FROM V where min(foo) >= 10
```

Even worse, the syntax you are using works fine if you have an index defined, but it works only because of an internal detail: OrientDB indexes defined on an collection attribute create an index entry for each element in the list. If you try to drop the index, you'll see that your query does not work anymore. IMHO this is not correct, an index should be an optimization, not a way to force a behavior.

I hope my point of view is clearer now

Thanks

Luigi

**Comment by T55134** [comment]
OK, let me see if I got you:
`SELECT FROM V where foo IN [1,2,3] ` 
**is working now** whether 'foo' is a single number or an embedded-list-long (on a list - it return true wether even ONE of the values if 'foo' exists inside the [1,2,3] parmeter list). 
According to what you wrote I should change it when 'foo' is an embedded-list' to:
`SELECT FROM V where foo CONTAINS 1 OR SELECT FROM V where foo CONTAINS 2 OR SELECT FROM V where foo CONTAINS 3`

Is that so? Sould I change it to this long way?
